tsdb/wal/wal_test.go

```diff
@@ -484,3 +484,32 @@ func BenchmarkWAL_Log(b *testing.B) {
 		})
 	}
 }
+
+// TestCloseGoroutineLeak detect goroutine leak when call Close of wal instance.
+func TestCloseGoroutineLeak(t *testing.T) {
+	dir, err := ioutil.TempDir("", "wal_close_goroutine_leak")
+	testutil.Ok(t, err)
+	defer func() {
+		testutil.Ok(t, os.RemoveAll(dir))
+	}()
+	w, err := NewSize(nil, nil, dir, pageSize, false)
+	testutil.Ok(t, err)
+
+	closedSegment, err := CreateSegment(dir, 0)
+	testutil.Ok(t, err)
+	err = w.setSegment(closedSegment)
+	testutil.Ok(t, err)
+
+	recordSize := (pageSize / 2) - recordHeaderSize
+	buf := make([]byte, recordSize)
+	_, err = rand.Read(buf)
+	testutil.Ok(t, err)
+
+	err = w.Log(buf)
+	testutil.Ok(t, err)
+
+	err = closedSegment.Close()
```

Comment:
We can add a comment here saying `Closing the segment before closing the WAL can cause goroutine leak. See PR description in https://github.com/prometheus/prometheus/pull/7655.`